### PR TITLE
Add `--gpu` option to the create command for NVIDIA support

### PR DESCRIPTION
**Pull request by misohu**

### CODE REVIEW

File: src/dss/main.py [hunk 1]
```diff
@@ -65,7 +65,15 @@ def initialize_command(kubeconfig: str) -> None:
     "--kubeconfig",
     help=f"Path to a Kubernetes config file. Defaults to the value of the KUBECONFIG environment variable, else to '{KUBECONFIG_DEFAULT}'.",  # noqa E501
 )
-def create_notebook_command(name: str, image: str, kubeconfig: str) -> None:
+@click.option("--no-gpu", is_flag=True, help="Create a notebook without GPU support.")
+@click.option(
+    "--gpu",
+    type=click.Choice(SUPPORTED_GPUS),
```
Comment (ca-scribner) [inline src/dss/main.py:71]: Does this need to be case sensitive?
```suggestion
    type=click.Choice(SUPPORTED_GPUS, case_sensitive=False),
```

File: src/dss/main.py [hunk 2]
```diff
@@ -65,7 +65,15 @@ def initialize_command(kubeconfig: str) -> None:
     "--kubeconfig",
     help=f"Path to a Kubernetes config file. Defaults to the value of the KUBECONFIG environment variable, else to '{KUBECONFIG_DEFAULT}'.",  # noqa E501
 )
-def create_notebook_command(name: str, image: str, kubeconfig: str) -> None:
+@click.option("--no-gpu", is_flag=True, help="Create a notebook without GPU support.")
```
Comment (ca-scribner) [inline src/dss/main.py:68]: Why do we need a `--no-gpu`?  Can the absence of `--gpu=something` imply no gpu, and we get rid of this?

File: src/dss/manifest_templates/notebook_deployment.yaml.j2 [hunk 3]
```diff
@@ -23,6 +23,11 @@ spec:
         - env:
           - name: MLFLOW_TRACKING_URI
             value: {{ mlflow_tracking_uri }}
+          {% if gpu %}
+          resources:
+            limits:
+              {{ gpu }}: 1
+          {% endif %}
```
Comment (ca-scribner) [inline src/dss/manifest_templates/notebook_deployment.yaml.j2:30]: Do we support only a single GPU per notebook?  If I have two GPUs, should we support using both?

File: src/dss/utils.py [hunk 4]
```diff
@@ -32,58 +32,88 @@ def __init__(self, msg: str, *args):
         self.msg = str(msg)
 
 
+def node_has_gpu_labels(lightkube_client: Client, labels: List[str]) -> bool:
```
Comment (ca-scribner) [inline src/dss/utils.py:35]: This asserts that all nodes have the provided labels, not that a given node has gpu labels.

Alternatively you could include the gpu labels here in the function code drop the `labels` input

```suggestion
def all_nodes_have_labels(lightkube_client: Client, labels: List[str]) -> bool:
```

File: src/dss/create_notebook.py [hunk 5]
```diff
@@ -78,51 +93,69 @@ def create_notebook(name: str, image: str, lightkube_client: Client) -> None:
...
         k8s_resource_handler.apply()
-
-        wait_for_deployment_ready(lightkube_client, namespace=DSS_NAMESPACE, deployment_name=name)
-
+        wait_for_deployment_ready(
+            lightkube_client, namespace=DSS_NAMESPACE, deployment_name=name, timeout_seconds=None
+        )
         logger.info(f"Success: Notebook {name} created successfully.")
+        if gpu:
+            logger.info(f"{gpu.title()} GPU attached to notebook.")
```
Comment (ca-scribner) [inline src/dss/create_notebook.py:101]: Minor suggestion.  `title()` wont be correct in all cases (ex: `Amd`).  I'd stick to the enumerators we say `gpu` should be 
```suggestion
            logger.info(f"{gpu} GPU attached to notebook.")
```

File: src/dss/create_notebook.py
Comment (ca-scribner) [inline src/dss/create_notebook.py:1]: I don't think the gpu requests are working as expected.  Try this:

```
dss initialize --kubeconfig ~/.kube/config 
dss create gpu-omitted --image=kubeflownotebookswg/jupyter-pytorch-cuda-full:latest --kubeconfig ~/.kube/config
dss create gpu-selected --image=kubeflownotebookswg/jupyter-pytorch-cuda-full:latest --kubeconfig ~/.kube/config
```

Then in each notebook server:
* create a terminal and do `nvidia-smi`.  The GPU should be visible only in one notebook server, but it is visible in both
* create a notebook and run `import torch; torch.cuda.is_available()`.  The GPU should be available only in one notebook server, but it is available in both
* create a notebook and run [this tutorial](https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html).  Both will run on the GPU.  After doing this, run `nvidia-smi` on the host system and we'll see two processes both using the GPU, like:
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.161.08             Driver Version: 535.161.08   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       On  | 00000000:00:1E.0 Off |                    0 |
| N/A   37C    P0              36W /  70W |    280MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A     70671      C   /opt/conda/bin/python                       158MiB |
|    0   N/A  N/A     96171      C   /opt/conda/bin/python                       118MiB |
+---------------------------------------------------------------------------------------+
```

Comment (ca-scribner) [reply src/dss/create_notebook.py:1]: One thing that is working correctly is the GPU requests.  My machine has one gpu, and if I do:
```
dss create x ... --gpu=nvidia
dss create y ... --gpu=nvidia
```

notebook `y` sits pending with the FailedScheduling warning: 1 Insufficient nvidia.com/gpu

**Review by ca-scribner — CHANGES_REQUESTED** [review]
see above comments

Tested by using an `g4dn.xlarge` ec2 instance in AWS.  Had to bump the storage to ~150GB to run multiple notebooks.  VM was set up with:
```
sudo snap install microk8s --channel 1.28/stable --classic
sudo usermod -a -G microk8s ubuntu
mkdir ~/.kube
sudo chown -R ubuntu ~/.kube
newgrp microk8s

# This starts a new terminal

microk8s enable storage dns rbac gpu 
microk8s config > ~/.kube/config
git clone http://github.com/canonical/data-science-stack
cd data-science-stack/
git checkout --track origin/KF-5420-add-gpu-flag-for-create
pip install -e .
python --version
nvidia-smi 
dss initialize --kubeconfig ~/.kube/config 
dss create nothing-specified --image=kubeflownotebookswg/jupyter-pytorch-cuda-full:latest --kubeconfig ~/.kube/config
dss create with-gpu --image=kubeflownotebookswg/jupyter-pytorch-cuda-full:latest --kubeconfig ~/.kube/config --gpu=nvidia
```

From there, I made sample notebooks out of [this pytorch example](https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html) to test if the gpu was working

File: src/dss/main.py (hunk 2 above)
Comment (misohu) [reply src/dss/main.py:68]: Its recommendation of ux team to provide this `--no-gpu` flag

Comment (ca-scribner) [reply src/dss/main.py:68]: I wonder if there's a miscommunication here.  If I understand correctly, these two commands are both the same?

```
dss create my-notebook
dss create my-notebook --no-gpu
```

Am I missing something about the CLI?
